Oregonator and Lotkavolterra_Oscillator models fail to run with C solvers

## Oregonator Error
```
Error encountered during execution of Makefile target: 'ode'.
Return code: 2- stdout: make: Entering directory '/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base'
g++ -std=c++14 -Wall -O3 -c -o /tmp/tmpr3_f3l9w/obj/ODESimulation.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESimulation.cpp -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include -I/tmp/tmpr3_f3l9w/template
g++ -std=c++14 -Wall -O3 -c -o /tmp/tmpr3_f3l9w/obj/ODESolver.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESolver.cpp  -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include -I/tmp/tmpr3_f3l9w/template
g++ -c -o /tmp/tmpr3_f3l9w/obj/cvode_nls.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_nls.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/cvode_io.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_io.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sundials_iterative.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_iterative.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/cvode_proj.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_proj.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sundials_matrix.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_matrix.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sunmatrix_band.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sunmatrix_band.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sunmatrix_dense.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sunmatrix_dense.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/cvode_ls.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_ls.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sundials_linearsolver.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_linearsolver.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sundials_nonlinearsolver.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_nonlinearsolver.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sundials_nvector_senswrapper.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_nvector_senswrapper.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sunnonlinsol_newton.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sunnonlinsol_newton.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sundials_nvector.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_nvector.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/nvector_serial.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/nvector_serial.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/cvode.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/cvode_spils.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_spils.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sundials_math.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_math.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmpr3_f3l9w/obj/sunlinsol_spgmr.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sunlinsol_spgmr.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -std=c++14 -Wall -O3 -c -o /tmp/tmpr3_f3l9w/obj/model.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base
g++ -std=c++14 -Wall -O3 -c -o /tmp/tmpr3_f3l9w/obj/arg_parser.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/arg_parser.cpp -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include -I/tmp/tmpr3_f3l9w/template
g++ -std=c++14 -Wall -O3 -o /tmp/tmpr3_f3l9w/Simulation.out /tmp/tmpr3_f3l9w/template/template.cpp /tmp/tmpr3_f3l9w/obj/model.o /tmp/tmpr3_f3l9w/obj/arg_parser.o -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include -I/tmp/tmpr3_f3l9w/template /tmp/tmpr3_f3l9w/obj/cvode_nls.o /tmp/tmpr3_f3l9w/obj/cvode_io.o /tmp/tmpr3_f3l9w/obj/sundials_iterative.o /tmp/tmpr3_f3l9w/obj/cvode_proj.o /tmp/tmpr3_f3l9w/obj/sundials_matrix.o /tmp/tmpr3_f3l9w/obj/sunmatrix_band.o /tmp/tmpr3_f3l9w/obj/sunmatrix_dense.o /tmp/tmpr3_f3l9w/obj/cvode_ls.o /tmp/tmpr3_f3l9w/obj/sundials_linearsolver.o /tmp/tmpr3_f3l9w/obj/sundials_nonlinearsolver.o /tmp/tmpr3_f3l9w/obj/sundials_nvector_senswrapper.o /tmp/tmpr3_f3l9w/obj/sunnonlinsol_newton.o /tmp/tmpr3_f3l9w/obj/sundials_nvector.o /tmp/tmpr3_f3l9w/obj/nvector_serial.o /tmp/tmpr3_f3l9w/obj/cvode.o /tmp/tmpr3_f3l9w/obj/cvode_spils.o /tmp/tmpr3_f3l9w/obj/sundials_math.o /tmp/tmpr3_f3l9w/obj/sunlinsol_spgmr.o /tmp/tmpr3_f3l9w/obj/ODESimulation.o /tmp/tmpr3_f3l9w/obj/ODESolver.o
make: Leaving directory '/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base'

- stderr: /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESolver.cpp: In function ‘void Gillespy::ODESolver(Gillespy::Simulation<double>*, double)’:
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESolver.cpp:49:7: warning: variable ‘flag’ set but not used [-Wunused-but-set-variable]
   49 |   int flag;
      |       ^~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp: In instantiation of ‘void Gillespy::Simulation<PType>::output_results_buffer(std::ostream&) [with PType = double; std::ostream = std::basic_ostream<char>]’:
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:141:18:   required from here
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:128:39: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  128 |   for (int trajectory = 0; trajectory < number_trajectories; trajectory++) {
      |                            ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:129:36: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  129 |    for (int timestep = 0; timestep < number_timesteps; timestep++) {
      |                           ~~~~~~~~~^~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:132:35: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  132 |     for (int species = 0; species < model->number_species; species++) {
      |                           ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp: In instantiation of ‘void Gillespy::Simulation<PType>::output_results_buffer(std::ostream&) [with PType = unsigned int; std::ostream = std::basic_ostream<char>]’:
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:142:18:   required from here
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:128:39: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  128 |   for (int trajectory = 0; trajectory < number_trajectories; trajectory++) {
      |                            ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:129:36: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  129 |    for (int timestep = 0; timestep < number_timesteps; timestep++) {
      |                           ~~~~~~~~~^~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:132:35: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  132 |     for (int species = 0; species < model->number_species; species++) {
      |                           ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/tmpr3_f3l9w/template/template.cpp:25:
/tmp/tmpr3_f3l9w/template/template_definitions.h:2:57: error: narrowing conversion of ‘2.0e+0’ from ‘double’ to ‘unsigned int’ [-Wnarrowing]
    2 | #define GPY_INIT_POPULATIONS {2.0,250.0,500.0,1000.0,0.0}
      |                                                         ^
/tmp/tmpr3_f3l9w/template/template.cpp:30:46: note: in expansion of macro ‘GPY_INIT_POPULATIONS’
   30 |  unsigned int populations[GPY_NUM_SPECIES] = GPY_INIT_POPULATIONS;
      |                                              ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template_definitions.h:2:57: error: narrowing conversion of ‘2.5e+2’ from ‘double’ to ‘unsigned int’ [-Wnarrowing]
    2 | #define GPY_INIT_POPULATIONS {2.0,250.0,500.0,1000.0,0.0}
      |                                                         ^
/tmp/tmpr3_f3l9w/template/template.cpp:30:46: note: in expansion of macro ‘GPY_INIT_POPULATIONS’
   30 |  unsigned int populations[GPY_NUM_SPECIES] = GPY_INIT_POPULATIONS;
      |                                              ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template_definitions.h:2:57: error: narrowing conversion of ‘5.0e+2’ from ‘double’ to ‘unsigned int’ [-Wnarrowing]
    2 | #define GPY_INIT_POPULATIONS {2.0,250.0,500.0,1000.0,0.0}
      |                                                         ^
/tmp/tmpr3_f3l9w/template/template.cpp:30:46: note: in expansion of macro ‘GPY_INIT_POPULATIONS’
   30 |  unsigned int populations[GPY_NUM_SPECIES] = GPY_INIT_POPULATIONS;
      |                                              ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template_definitions.h:2:57: error: narrowing conversion of ‘1.0e+3’ from ‘double’ to ‘unsigned int’ [-Wnarrowing]
    2 | #define GPY_INIT_POPULATIONS {2.0,250.0,500.0,1000.0,0.0}
      |                                                         ^
/tmp/tmpr3_f3l9w/template/template.cpp:30:46: note: in expansion of macro ‘GPY_INIT_POPULATIONS’
   30 |  unsigned int populations[GPY_NUM_SPECIES] = GPY_INIT_POPULATIONS;
      |                                              ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘V’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmpr3_f3l9w/template/template_definitions.h:1:30: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,2.0) CONSTANT(P1,0.1) CONSTANT(P2,104.0) CONSTANT(P3,4e-07) CONSTANT(P4,26.0)
      |                              ^~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘P0’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmpr3_f3l9w/template/template_definitions.h:1:46: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,2.0) CONSTANT(P1,0.1) CONSTANT(P2,104.0) CONSTANT(P3,4e-07) CONSTANT(P4,26.0)
      |                                              ^~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘P1’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmpr3_f3l9w/template/template_definitions.h:1:63: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,2.0) CONSTANT(P1,0.1) CONSTANT(P2,104.0) CONSTANT(P3,4e-07) CONSTANT(P4,26.0)
      |                                                               ^~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘P2’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmpr3_f3l9w/template/template_definitions.h:1:80: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,2.0) CONSTANT(P1,0.1) CONSTANT(P2,104.0) CONSTANT(P3,4e-07) CONSTANT(P4,26.0)
      |                                                                                ^~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘P3’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmpr3_f3l9w/template/template_definitions.h:1:99: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,2.0) CONSTANT(P1,0.1) CONSTANT(P2,104.0) CONSTANT(P3,4e-07) CONSTANT(P4,26.0)
      |                                                                                                   ^~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘P4’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmpr3_f3l9w/template/template_definitions.h:1:118: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,2.0) CONSTANT(P1,0.1) CONSTANT(P2,104.0) CONSTANT(P3,4e-07) CONSTANT(P4,26.0)
      |                                                                                                                      ^~~~~~~~
/tmp/tmpr3_f3l9w/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
make: *** [/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Makefile:87: ode] Error 1
```

## Lotkavolterra_Oscillator Error
```
Error encountered during execution of Makefile target: 'ode'.
Return code: 2- stdout: make: Entering directory '/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base'
g++ -std=c++14 -Wall -O3 -c -o /tmp/tmp71yhqxkf/obj/ODESimulation.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESimulation.cpp -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include -I/tmp/tmp71yhqxkf/template
g++ -std=c++14 -Wall -O3 -c -o /tmp/tmp71yhqxkf/obj/ODESolver.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESolver.cpp  -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include -I/tmp/tmp71yhqxkf/template
g++ -c -o /tmp/tmp71yhqxkf/obj/cvode_nls.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_nls.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/cvode_io.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_io.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sundials_iterative.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_iterative.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/cvode_proj.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_proj.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sundials_matrix.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_matrix.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sunmatrix_band.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sunmatrix_band.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sunmatrix_dense.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sunmatrix_dense.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/cvode_ls.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_ls.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sundials_linearsolver.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_linearsolver.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sundials_nonlinearsolver.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_nonlinearsolver.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sundials_nvector_senswrapper.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_nvector_senswrapper.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sunnonlinsol_newton.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sunnonlinsol_newton.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sundials_nvector.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_nvector.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/nvector_serial.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/nvector_serial.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/cvode.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/cvode_spils.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/cvode_spils.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sundials_math.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sundials_math.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -c -o /tmp/tmp71yhqxkf/obj/sunlinsol_spgmr.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/src/sunlinsol_spgmr.c -Wall -O3 -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include
g++ -std=c++14 -Wall -O3 -c -o /tmp/tmp71yhqxkf/obj/model.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base
g++ -std=c++14 -Wall -O3 -c -o /tmp/tmp71yhqxkf/obj/arg_parser.o /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/arg_parser.cpp -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include -I/tmp/tmp71yhqxkf/template
g++ -std=c++14 -Wall -O3 -o /tmp/tmp71yhqxkf/Simulation.out /tmp/tmp71yhqxkf/template/template.cpp /tmp/tmp71yhqxkf/obj/model.o /tmp/tmp71yhqxkf/obj/arg_parser.o -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base -I/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Sundials/include -I/tmp/tmp71yhqxkf/template /tmp/tmp71yhqxkf/obj/cvode_nls.o /tmp/tmp71yhqxkf/obj/cvode_io.o /tmp/tmp71yhqxkf/obj/sundials_iterative.o /tmp/tmp71yhqxkf/obj/cvode_proj.o /tmp/tmp71yhqxkf/obj/sundials_matrix.o /tmp/tmp71yhqxkf/obj/sunmatrix_band.o /tmp/tmp71yhqxkf/obj/sunmatrix_dense.o /tmp/tmp71yhqxkf/obj/cvode_ls.o /tmp/tmp71yhqxkf/obj/sundials_linearsolver.o /tmp/tmp71yhqxkf/obj/sundials_nonlinearsolver.o /tmp/tmp71yhqxkf/obj/sundials_nvector_senswrapper.o /tmp/tmp71yhqxkf/obj/sunnonlinsol_newton.o /tmp/tmp71yhqxkf/obj/sundials_nvector.o /tmp/tmp71yhqxkf/obj/nvector_serial.o /tmp/tmp71yhqxkf/obj/cvode.o /tmp/tmp71yhqxkf/obj/cvode_spils.o /tmp/tmp71yhqxkf/obj/sundials_math.o /tmp/tmp71yhqxkf/obj/sunlinsol_spgmr.o /tmp/tmp71yhqxkf/obj/ODESimulation.o /tmp/tmp71yhqxkf/obj/ODESolver.o
make: Leaving directory '/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base'

- stderr: /opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESolver.cpp: In function ‘void Gillespy::ODESolver(Gillespy::Simulation<double>*, double)’:
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESolver.cpp:49:7: warning: variable ‘flag’ set but not used [-Wunused-but-set-variable]
   49 |   int flag;
      |       ^~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp: In instantiation of ‘void Gillespy::Simulation<PType>::output_results_buffer(std::ostream&) [with PType = double; std::ostream = std::basic_ostream<char>]’:
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:141:18:   required from here
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:128:39: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  128 |   for (int trajectory = 0; trajectory < number_trajectories; trajectory++) {
      |                            ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:129:36: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  129 |    for (int timestep = 0; timestep < number_timesteps; timestep++) {
      |                           ~~~~~~~~~^~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:132:35: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  132 |     for (int species = 0; species < model->number_species; species++) {
      |                           ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp: In instantiation of ‘void Gillespy::Simulation<PType>::output_results_buffer(std::ostream&) [with PType = unsigned int; std::ostream = std::basic_ostream<char>]’:
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:142:18:   required from here
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:128:39: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  128 |   for (int trajectory = 0; trajectory < number_trajectories; trajectory++) {
      |                            ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:129:36: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  129 |    for (int timestep = 0; timestep < number_timesteps; timestep++) {
      |                           ~~~~~~~~~^~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/model.cpp:132:35: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  132 |     for (int species = 0; species < model->number_species; species++) {
      |                           ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/tmp71yhqxkf/template/template.cpp:25:
/tmp/tmp71yhqxkf/template/template_definitions.h:2:43: error: narrowing conversion of ‘9.4999999999999998e-3’ from ‘double’ to ‘unsigned int’ [-Wnarrowing]
    2 | #define GPY_INIT_POPULATIONS {0.0095,0.008}
      |                                           ^
/tmp/tmp71yhqxkf/template/template.cpp:30:46: note: in expansion of macro ‘GPY_INIT_POPULATIONS’
   30 |  unsigned int populations[GPY_NUM_SPECIES] = GPY_INIT_POPULATIONS;
      |                                              ^~~~~~~~~~~~~~~~~~~~
/tmp/tmp71yhqxkf/template/template_definitions.h:2:43: error: narrowing conversion of ‘8.0000000000000002e-3’ from ‘double’ to ‘unsigned int’ [-Wnarrowing]
    2 | #define GPY_INIT_POPULATIONS {0.0095,0.008}
      |                                           ^
/tmp/tmp71yhqxkf/template/template.cpp:30:46: note: in expansion of macro ‘GPY_INIT_POPULATIONS’
   30 |  unsigned int populations[GPY_NUM_SPECIES] = GPY_INIT_POPULATIONS;
      |                                              ^~~~~~~~~~~~~~~~~~~~
/tmp/tmp71yhqxkf/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘V’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmp71yhqxkf/template/template_definitions.h:1:30: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,0.004) CONSTANT(P1,0.5) CONSTANT(P2,0.0045)
      |                              ^~~~~~~~
/tmp/tmp71yhqxkf/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
/tmp/tmp71yhqxkf/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘P0’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmp71yhqxkf/template/template_definitions.h:1:46: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,0.004) CONSTANT(P1,0.5) CONSTANT(P2,0.0045)
      |                                              ^~~~~~~~
/tmp/tmp71yhqxkf/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
/tmp/tmp71yhqxkf/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘P1’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmp71yhqxkf/template/template_definitions.h:1:65: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,0.004) CONSTANT(P1,0.5) CONSTANT(P2,0.0045)
      |                                                                 ^~~~~~~~
/tmp/tmp71yhqxkf/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
/tmp/tmp71yhqxkf/template/template.cpp:60:52: warning: unnecessary parentheses in declaration of ‘P2’ [-Wparentheses]
   60 |  #define CONSTANT(name, value) static const double (name) = (value);
      |                                                    ^
/tmp/tmp71yhqxkf/template/template_definitions.h:1:82: note: in expansion of macro ‘CONSTANT’
    1 | #define GPY_PARAMETER_VALUES CONSTANT(V,1.0) CONSTANT(P0,0.004) CONSTANT(P1,0.5) CONSTANT(P2,0.0045)
      |                                                                                  ^~~~~~~~
/tmp/tmp71yhqxkf/template/template.cpp:61:2: note: in expansion of macro ‘GPY_PARAMETER_VALUES’
   61 |  GPY_PARAMETER_VALUES
      |  ^~~~~~~~~~~~~~~~~~~~
make: *** [/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_base/Makefile:87: ode] Error 1
```

--- a/gillespy2/solvers/cpp/c_base/model.cpp
+++ b/gillespy2/solvers/cpp/c_base/model.cpp
@@ -19,19 +19,21 @@
 #include "model.h"
 
 namespace Gillespy {
-	Model::Model(
+
+	template <typename PType>
+	Model<PType>::Model(
 		std::vector<std::string> species_names,
-		std::vector<unsigned int> species_populations,
+		std::vector<double> species_populations,
 		std::vector<std::string> reaction_names) :
 		number_species(species_names.size()),
 		number_reactions(reaction_names.size()) {
 
-		species = std::make_unique<Species[]>(number_species);
+		species = std::make_unique<Species<PType>[]>(number_species);
 		reactions = std::make_unique<Reaction[]>(number_reactions);
 
 		for (unsigned int i = 0; i < number_species; i++) {
 			species[i].id = i;
-			species[i].initial_population = species_populations[i];
+			species[i].initial_population = static_cast<PType>(species_populations[i]);
 			species[i].name = species_names[i];
 		}
 
@@ -47,7 +49,8 @@ namespace Gillespy {
 		}
 	}
 
-	void Model::update_affected_reactions() {
+	template <typename PType>
+	void Model<PType>::update_affected_reactions() {
 		// Clear affected_reactions for each reaction.
 		for (unsigned int i = 0; i < number_reactions; i++) {
 			reactions[i].affected_reactions.clear();
@@ -66,7 +69,7 @@ namespace Gillespy {
 	}
 
 	template <typename TNum>
-	void init_timeline(Model *model, Simulation<TNum> &simulation) {
+	void init_timeline(Model<TNum> *model, Simulation<TNum> &simulation) {
 		double timestep_size = simulation.end_time / (simulation.number_timesteps - 1);
 		simulation.timeline = new double[simulation.number_timesteps];
 
@@ -76,7 +79,7 @@ namespace Gillespy {
 	}
 
 	template <typename TNum>
-	void init_simulation(Model *model, Simulation<TNum> &simulation) {
+	void init_simulation(Model<TNum> *model, Simulation<TNum> &simulation) {
 		init_timeline(model, simulation);
 
 		unsigned int trajectory_size = simulation.number_timesteps * (model->number_species);
@@ -138,9 +141,16 @@ namespace Gillespy {
 		os << (int)current_time;
 	}
 
+	// DETERMINISTIC SIMULATIONS: explicit instantation of real-valued data structures.
+	template struct Species<double>;
+	template struct Model<double>;
 	template struct Simulation<double>;
+
+	// STOCHASTIC SIMULATIONS: explicit instantiation of discrete-valued data structures.
+	template struct Species<unsigned int>;
+	template struct Model<unsigned int>;
 	template struct Simulation<unsigned int>;
 
-	template void init_simulation<double>(Model *model, Simulation<double> &simulation);
-	template void init_simulation<unsigned int>(Model *model, Simulation<unsigned int> &simulation);
+	template void init_simulation<double>(Model<double> *model, Simulation<double> &simulation);
+	template void init_simulation<unsigned int>(Model<unsigned int> *model, Simulation<unsigned int> &simulation);
 }

--- a/gillespy2/solvers/cpp/c_base/model.h
+++ b/gillespy2/solvers/cpp/c_base/model.h
@@ -25,9 +25,11 @@
 #include <iostream>
 
 namespace Gillespy {
+
+	template <typename PType>
 	struct Species {
 		unsigned int id;
-		unsigned int initial_population;
+		PType initial_population;
 
 		std::string name;
 
@@ -48,18 +50,19 @@ namespace Gillespy {
 		std::unique_ptr<int[]> species_change;
 	};
 
+	template <typename PType>
 	struct Model {
 		unsigned int number_species;
 		unsigned int number_reactions;
 
-		std::unique_ptr<Species[]> species;
+		std::unique_ptr<Species<PType>[]> species;
 		std::unique_ptr<Reaction[]> reactions;
 
 		void update_affected_reactions();
 
 		Model(
 			std::vector<std::string> species_names,
-			std::vector<unsigned int> species_populations,
+			std::vector<double> species_populations,
 			std::vector<std::string> reaction_names
 		);
 	};
@@ -88,7 +91,7 @@ namespace Gillespy {
 		PType *trajectories_1D;
 		PType ***trajectories;
 
-		Model *model;
+		Model<PType> *model;
 
 		IPropensityFunction *propensity_function;
 
@@ -99,11 +102,29 @@ namespace Gillespy {
 	};
 
 	template <typename TNum>
-	void init_simulation(Model *model, Simulation<TNum> &simulation);
+	void init_simulation(Model<TNum> *model, Simulation<TNum> &simulation);
 
+	/* ================================= *
+	 * === DETERMINISTIC SIMULATIONS === *
+	 * ================================= */
+
+	// Deterministic Species: species whose initial and runtime values are continuous.
+	extern template struct Species<double>;
+	// Deterministic Model: species state represented using concentration values.
+	extern template struct Model<double>;
+	// Deterministic Simulation: run using a continuous-valued model.
 	extern template struct Simulation<double>;
-	extern template void init_simulation<double>(Model *model, Simulation<double> &simulation);
+	extern template void init_simulation<double>(Model<double> *model, Simulation<double> &simulation);
 
+	/* ============================== *
+	 * === STOCHASTIC SIMULATIONS === *
+	 * ============================== */
+
+	// Stochastic Species: species whose initial and runtime values are discrete.
+	extern template struct Species<unsigned int>;
+	// Stochastic Model: species state represented using discrete population values.
+	extern template struct Model<unsigned int>;
+	// Stochastic Simulation: run using a discrete-valued model.
 	extern template struct Simulation<unsigned int>;
-	extern template void init_simulation<unsigned int>(Model *model, Simulation<unsigned int> &simulation);
+	extern template void init_simulation<unsigned int>(Model<unsigned int> *model, Simulation<unsigned int> &simulation);
 }

--- a/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESimulation.cpp
+++ b/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESimulation.cpp
@@ -72,7 +72,7 @@ int main(int argc, char *argv[]) {
 	number_timesteps = parser.timesteps;
 	increment = parser.increment;
 
-	Model model(species_names, species_populations, reaction_names);
+	Model<double> model(species_names, species_populations, reaction_names);
 	add_reactions(model);
 
 	if (seed_time)

--- a/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESolver.cpp
+++ b/gillespy2/solvers/cpp/c_base/ode_cpp_solver/ODESolver.cpp
@@ -60,7 +60,7 @@ namespace Gillespy
 		sunindextype N = (simulation->model)->number_species;
 		N_Vector y0 = N_VNew_Serial(N);
 
-		Model *simulation_model = simulation->model;
+		Model<double> *simulation_model = simulation->model;
 
 		// Add species initial conditions to the current state vectory `y0`.
 		for (unsigned int species_index = 0; species_index < simulation_model->number_species; species_index++)
@@ -138,7 +138,7 @@ namespace Gillespy
 	{
 		UserData *sim_data = (UserData *)user_data;
 		Simulation<double> *simulation = sim_data->my_sim;
-		Model *model = simulation->model;
+		Model<double> *model = simulation->model;
 
 		// N_VGetArrayPointer returns a pointer to the data property within N_Vector.
 		realtype *ydata = N_VGetArrayPointer(y);

--- a/gillespy2/solvers/cpp/c_base/ssa_cpp_solver/SSASimulation.cpp
+++ b/gillespy2/solvers/cpp/c_base/ssa_cpp_solver/SSASimulation.cpp
@@ -72,7 +72,7 @@ int main(int argc, char *argv[])
 	number_trajectories = parser.trajectories;
 	number_timesteps = parser.timesteps;
 	
-	Model model(species_names, species_populations, reaction_names);
+	Model<unsigned int> model(species_names, species_populations, reaction_names);
 	add_reactions(model);
 	
 	if (seed_time)    

--- a/gillespy2/solvers/cpp/c_base/tau_leaping_cpp_solver/TauLeapingSimulation.cpp
+++ b/gillespy2/solvers/cpp/c_base/tau_leaping_cpp_solver/TauLeapingSimulation.cpp
@@ -73,7 +73,7 @@ int main(int argc, char* argv[]){
 	number_timesteps = parser.timesteps;
 	tau_tol = parser.tau_tol;
 
-	Model model(species_names, species_populations, reaction_names);
+	Model<unsigned int> model(species_names, species_populations, reaction_names);
 	add_reactions(model);
 
 	if(seed_time) {

--- a/gillespy2/solvers/cpp/c_base/tau_leaping_cpp_solver/TauLeapingSolver.cpp
+++ b/gillespy2/solvers/cpp/c_base/tau_leaping_cpp_solver/TauLeapingSolver.cpp
@@ -45,7 +45,7 @@ namespace Gillespy
 		int critical_threshold = 10;
 
 		std::map<std::string, int> HOR;
-		std::set<Gillespy::Species> reactants;
+		std::set<Gillespy::Species<unsigned int>> reactants;
 
 		std::map<std::string, std::function<double(double)>> g_i_lambdas;
 		std::map<std::string, int> g_i;
@@ -54,7 +54,7 @@ namespace Gillespy
 		std::map<int, std::vector<int>> products;
 	};
 
-	TauArgs initialize(Gillespy::Model &model, double tau_tol)
+	TauArgs initialize(Gillespy::Model<unsigned int> &model, double tau_tol)
 	{
 		// Initialize TauArgs struct to be returned as a pointer
 		TauArgs tau_args;
@@ -139,7 +139,7 @@ namespace Gillespy
 	}
 
 	double select(
-		Gillespy::Model &model,
+		Gillespy::Model<unsigned int> &model,
 		TauArgs &tau_args,
 		const double &tau_tol,
 		const double &current_time,
@@ -289,7 +289,7 @@ namespace Gillespy
 	}
 
 	std::pair<std::map<std::string, int>, double> get_reactions(
-		const Gillespy::Model *model,
+		const Gillespy::Model<unsigned int> *model,
 		const std::vector<double> &propensity_values,
 		double tau_step,
 		double current_time,

--- a/gillespy2/solvers/cpp/c_base/template/template.cpp
+++ b/gillespy2/solvers/cpp/c_base/template/template.cpp
@@ -27,10 +27,10 @@
 
 namespace Gillespy
 {
-	unsigned int populations[GPY_NUM_SPECIES] = GPY_INIT_POPULATIONS;
-	std::vector<unsigned int> species_populations(
+	double populations[GPY_NUM_SPECIES] = GPY_INIT_POPULATIONS;
+	std::vector<double> species_populations(
 		populations,
-		populations + sizeof(populations) / sizeof(unsigned int));
+		populations + sizeof(populations) / sizeof(double));
 
 	std::string s_names[GPY_NUM_SPECIES] = 
 	{
@@ -118,7 +118,8 @@ namespace Gillespy
 		}
 	}
 
-	void add_reactions(Model &model)
+	template <typename T>
+	void add_reactions(Model<T> &model)
 	{
 		unsigned int rxn_i;
 		unsigned int spec_i;
@@ -138,4 +139,7 @@ namespace Gillespy
 
 		model.update_affected_reactions();
 	}
+
+	template void add_reactions<double>(Model<double> &model);
+	template void add_reactions<unsigned int>(Model<unsigned int> &model);
 }

--- a/gillespy2/solvers/cpp/c_base/template/template.h
+++ b/gillespy2/solvers/cpp/c_base/template/template.h
@@ -26,15 +26,20 @@
 
 namespace Gillespy
 {
-	extern std::vector<unsigned int> species_populations;
+	extern std::vector<double> species_populations;
 	extern std::vector<std::string> species_names;
 	extern std::vector<std::string> reaction_names;
 
 	double map_propensity(int reaction_id, const std::vector<int> &state);
 	double map_propensity(int reaction_id, unsigned int *S);
 	double map_ode_propensity(int reaction_id, const std::vector<double> &state);
-	void add_reactions(Model &model);
+
+	template <typename T>
+	void add_reactions(Model<T> &model);
 
 	void map_variable_parameters(std::stringstream &stream);
 	void map_variable_populations(std::stringstream &stream);
+
+	extern template void add_reactions<double>(Model<double> &model);
+	extern template void add_reactions<unsigned int>(Model<unsigned int> &model);
 }

